src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -133,7 +134,77 @@ internal void SendPendingBatches()
             // Dispatch any buffered renders we accumulated during a disconnect.
             // Note that while the rendering is async, we cannot await it here. The Task returned by ProcessBufferedRenderBatches relies on
             // OnRenderCompleted to be invoked to complete, and SignalR does not allow concurrent hub method invocations.
-            var _ = Renderer.Dispatcher.InvokeAsync(() => Renderer.ProcessBufferedRenderBatches());
+            _ = Renderer.Dispatcher.InvokeAsync(() => Renderer.ProcessBufferedRenderBatches());
+        }
+
+        public async Task EndInvokeJSFromDotNet(long asyncCall, bool succeded, string arguments)
+        {
+            try
+            {
+                AssertInitialized();
+
+                await Renderer.Dispatcher.InvokeAsync(() =>
+                {
+                    SetCurrentCircuitHost(this);
+                    if (!succeded)
+                    {
+                        // We can log the arguments here because it is simply the JS error with the call stack.
+                        Log.EndInvokeJSFailed(_logger, asyncCall, arguments);
+                    }
+                    else
+                    {
+                        Log.EndInvokeJSSucceeded(_logger, asyncCall);
+                    }
+
+                    DotNetDispatcher.EndInvoke(arguments);
+                });
+            }
+            catch (Exception ex)
+            {
+                Log.EndInvokeDispatchException(_logger, ex);
+            }
+        }
+
+        public async Task DispatchEvent(string eventDescriptorJson, string eventArgs)
+        {
+            RendererRegistryEventDispatcher.BrowserEventDescriptor eventDescriptor = null;
+            try
+            {
+                AssertInitialized();
+                eventDescriptor = ParseEventDescriptor(eventDescriptorJson);
+
+                await Renderer.Dispatcher.InvokeAsync(() =>
+                {
+                    SetCurrentCircuitHost(this);
+
+                    if (eventDescriptor != null)
+                    {
+                        return RendererRegistryEventDispatcher.DispatchEvent(eventDescriptor, eventArgs);
+                    }
```

Comment:
The logic seems a little odd. If we know `eventDescriptor` is `null`, why do we go through the block from L176 to L186? Couldn't we just `return` after L174 if `eventDescriptor` is `null`?

Reply:
That's fair, I don't know what I was thinking about. I'll update it.

Reply:
done